nemo/collections/nlp/models/nlp_model.py

```diff
@@ -80,44 +81,66 @@ def register_bert_model(self):
         """Instantiates tokenizer based on config and registers tokenizer artifacts.
 
         Args:
             cfg (DictConfig): Tokenizer config
         """
+        if self._is_model_being_restored() and os.path.exists('tokenizer.vocab_file'):
+            # model is being restored from .nemo file so tokenizer.vocab_file has precedence
+            vocab_file = self.register_artifact(config_path='tokenizer.vocab_file', src='tokenizer.vocab_file')
+        elif cfg.vocab_file is not None:
+            # use vocab file from config
+            vocab_file = self.register_artifact(config_path='tokenizer.vocab_file', src=cfg.vocab_file)
+        else:
+            vocab_file = None
+
         tokenizer = get_tokenizer(
             tokenizer_name=cfg.tokenizer_name,
-            vocab_file=cfg.vocab_file,
+            vocab_file=vocab_file,
             special_tokens=OmegaConf.to_container(cfg.special_tokens) if cfg.special_tokens else None,
-            tokenizer_model=cfg.tokenizer_model,
+            tokenizer_model=self.register_artifact(config_path='tokenizer.tokenizer_model', src=cfg.tokenizer_model),
         )
         self.tokenizer = tokenizer
-        self.register_tokenizer(cfg)
+
+        if vocab_file is None:
+            # when there is no vocab file we try to get the vocab from the tokenizer and register it
+            self._register_vocab_from_tokenizer(config_path='tokenizer.vocab_file')
 
     @rank_zero_only
-    def register_tokenizer(self, cfg: DictConfig):
-        """Adds tokenizer vocab file and model to .nemo archive.
+    def _register_vocab_from_tokenizer(self, config_path='tokenizer.vocab_file'):
+        """Creates vocab file from tokenizer if vocab file is None.
 
         Args:
-            cfg (DictConfig): Tokenizer config.
+            src (str): Path to vocab file.
         """
         vocab_file_config_path = 'tokenizer.vocab_file'
         vocab_dict_config_path = 'tokenizer_vocab_dict.json'
         if self.tokenizer is None:
-            raise ValueError('Instantiate self.tokenizer before registering it.')
+            raise ValueError('Instantiate self.tokenizer before registering vocab from it.')
         else:
-            if cfg.vocab_file is not None:
-                self.register_artifact(config_path=vocab_file_config_path, src=cfg.vocab_file)
-            elif isinstance(self.tokenizer, AutoTokenizer):
+            if isinstance(self.tokenizer, AutoTokenizer):
                 # extract vocab from tokenizer
-                vocab_json_src = os.path.join(NEMO_NLP_TMP, vocab_dict_config_path)
                 vocab_dict = self.tokenizer.tokenizer.get_vocab()
+
+                # get hash of vocab_dict to create a unique directory to write vocab_dict and vocab_file
+                m = hashlib.md5()
+                # get string representation of vocab_dict
+                vocab_dict_str = json.dumps(vocab_dict, sort_keys=True).encode()
+                m.update(vocab_dict_str)
```

Comment:
This will fail for tokekizer name is bert-base-case and bert-base-uncased with exact same vocabulary content.  The likelihood of such an occurrence is low but possible, so let's do a hashlib.updade() before this with the value of cfg.tokenizer_name. 

That way even if vocab is exact match, as long as the Tokenizers name is also different, different hash directory should be created.

Reply:
Good idea, added it.